ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/recovery/RecoveryModule.java

```diff
@@ -34,4 +34,13 @@ public interface RecoveryModule
      */
 
     public void periodicWorkSecondPass ();
+
+    /**
+     * Called by the RecoveryManager after the completion of the second pass.
+     *
+     * @return true if there is work left to do, false otherwise
+     */
+    default boolean hasWork() {
```

Comment:
@tomjenkinson 

This matter has been discussed internally, and we have reached an agreement. I will now resolve this conversation.

Reply:
This is an open source project, will you share the rationale with the community.

Reply:
I have had a look into this and there is an important distinction between ExtendedRecoveryModule and the new default method in RecoveryModule.

The ExtendedRecoveryModule interface allows the programmatic detection of general problems in periodic recovery, which is a slightly different feature than this new graceful shutdown functionality can use which is that we only want to know from recovery modules if there is work in the module which the recovery manager can still perform.

As an example, we want to optionally not prevent heuristic issues from blocking shutdown. Please consider that the way ExtendedRecoveryModule has been implemented by XARecoveryModule I can see at least one example where a problem related to a heuristic would result in a boolean response from the ERM that can be anticipated that (if it were integrated into our new graceful shutdown logic) it would prevent our graceful shutdown for completing (until an administrator resolved the heuristic). (For reference, the example is https://github.com/jbosstm/narayana/blob/6798bccf35309c42fbc2889e0fe14f722331cdcb/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java#L944)